src/app/public/modules/datepicker/fuzzy-date.service.ts

```diff
@@ -0,0 +1,287 @@
+
+    let year: any;
+
+    // Find the number value in the string that is 4 digits long.
+    dateString.split(separator).forEach(dateComponent => {
+      if (!year && parseInt(dateComponent, 10).toString().length === 4) {
+        year = dateComponent;
+      }
+    });
+
+    if (year && !isNaN(year)) {
+        return parseInt(year, 10);
+    }
+  }
+
+  public getMomentFromFuzzyDate(fuzzyDate: any): any {
+    if (!fuzzyDate) { return; }
+
+    // Year to use for fuzzy dates that don't have a year.
+    let defaultYear: number = new Date().getFullYear();
+
+    if (fuzzyDate.month === 2 && fuzzyDate.day === 29) {
+      // Needs to be a valid year for leap years.
+      defaultYear = this.getMostRecentLeapYear(defaultYear);
+    }
+
+    return moment([fuzzyDate.year || defaultYear, (fuzzyDate.month - 1) || 0, fuzzyDate.day || 1]);
+  }
+
+  public getDateStringFromFuzzyDate(fuzzyDate: any, dateFormatString: any): string {
+    if (!fuzzyDate || !dateFormatString) { return; }
+
+    let dateString: string = '',
+        separator: any,
+        dateFormatIndexes = this.getDateFormatIndexes(dateFormatString);
+
+    // Get the components of the date in the order expected of the local format
+    let dateComponents = [
+        { value: fuzzyDate.year || 0, index: dateFormatIndexes.yearIndex },
+        { value: fuzzyDate.month || 0, index: dateFormatIndexes.monthIndex },
+        { value: fuzzyDate.day || 0, index: dateFormatIndexes.dayIndex }
+    ];
+    dateComponents.sort(function (a: any, b: any) { return a.index - b.index; });
+
+    separator = this.getSeparatorFromDateString(dateFormatString);
+
+    dateComponents.forEach(component => {
+      if (component.value > 0 && component.index > -1) {
+        if (dateString) {
+            dateString += separator;
+        }
+
+        dateString += component.value.toString();
+      }
+    });
+
+    return dateString.trim();
+  }
+
+  public getFuzzyDateFromSelectedDate(selectedDate: Date, dateFormatString: string): any {
+    if (!selectedDate || !dateFormatString) { return; }
+
+    let fuzzyDate: SkyFuzzyDate = {},
+        dateFormatIndexes = this.getDateFormatIndexes(dateFormatString);
+
+    if (dateFormatIndexes.yearIndex > -1) {
+      fuzzyDate.year = selectedDate.getFullYear();
+    }
+
+    if (dateFormatIndexes.dayIndex > -1) {
+      fuzzyDate.day = selectedDate.getDate();
+    }
+
+    if (dateFormatIndexes.monthIndex > -1) {
+      // getMonth returns month 0 through 11
+      fuzzyDate.month = selectedDate.getMonth() + 1;
+    }
+
+    return fuzzyDate;
+  }
+
+  public getFuzzyDateFromDateString(dateString: string, dateFormatString: string): any {
+    if (!dateString || !dateFormatString) { return; }
+
+    let day,
+        month,
+        year,
+        separator,
+        dateFormatIndexes = this.getDateFormatIndexes(dateFormatString),
+        dateComponents,
+        dateComponentIndexes = [],
+        yearIndex,
+        monthIndex,
+        dayIndex;
```

Comment:
This is a minor code style thing, but would you mind declaring each one of these properties with a type? We'll have to change those before we merge to production, so if you do this anywhere else, would you mind keeping those consistent for us? Thanks!
```
    let day: string;
    let month: string;
    let year: string;
```